### PR TITLE
Bug 1874322: add bcrypt as a supported hashing method for htpasswd passwords 

### PR DESCRIPTION
**Pull request by stlaz**
The PR is based on https://github.com/bitly/oauth2_proxy/commit/008ffae3bb5f1d068f2ea4dddb88ea80a6697297 with the expection of not including the `$2b$` prefix based on https://svn.apache.org/viewvc/apr/apr/trunk/crypto/crypt_blowfish.c?view=markup#l580

### CODE REVIEW

**Comment by stlaz** [comment]
/retest
impossibru, the unit test shouldn't fail

File: htpasswd.go [hunk 1]
```diff
@@ -47,14 +49,19 @@ func (h *HtpasswdFile) Validate(user string, password string) bool {
...
-			return true
-		}
-	} else {
-		log.Printf("Invalid htpasswd entry for %s. Must be a SHA entry.", user)
+		return shaValue == base64.StdEncoding.EncodeToString(d.Sum(nil))
+	}
+
+	bcryptPrefix := realPassword[:4]
+	if bcryptPrefix == "$2a$" || bcryptPrefix == "$2x$" || bcryptPrefix == "$2y$" {
+		return bcrypt.CompareHashAndPassword([]byte(realPassword), []byte(password)) == nil
```
Comment (deads2k) [inline htpasswd.go:62]: you intentionally pass along the `$2whatever`?  Also, where did *that* format come from?

**Comment by deads2k** [comment]
I don't think this will actually resolve the bug since the bug is about the possibility to store passwords in sha1, not the requirement to do so.

Also, why do we attempt to verify passwords in this proxy at all?  That doesn't make much sense to me.

File: htpasswd.go (hunk 1 above)
Comment (stlaz) [reply htpasswd.go:62]: this is what upstream of this project did before it went archived, the format comes from https://svn.apache.org/viewvc/apr/apr/trunk/crypto/crypt_blowfish.c?view=markup#l580

**Comment by stlaz** [comment]
@deads2k we can't remove the sha1-logins as that would break everyone who is using those, unfortunately.

This is not exactly password verification for the oauth-users, it is a htpasswd config that allows to bypass the oauth-login requirement. I think it might be though about as kind of a service-account idp for the proxy.

**Comment by stlaz** [comment]
/remove-lifecycle stale
for now

File: htpasswd.go (hunk 1 above)
Comment (s-urbaniak) [reply htpasswd.go:62]: suggestion: maybe let's add a comment to this external reference

**Comment by s-urbaniak** [comment]
agreed with @stlaz, this just adds the bcrypt capability, while not removing sha1

**Pull request merged**